### PR TITLE
149 course completion

### PR DESCRIPTION
**Pull request by koshalt**
Partially fixing #149 with open questions

### CODE REVIEW

File: mobile-academy/src/main/java/org/motechproject/nms/mobileacademy/domain/CompletionRecord.java [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+import org.motechproject.mds.annotations.Field;
+
+/**
+ * Tracks the completion record for a given calling number
+ */
+@Entity(tableName = "nms_ma_completion_records")
+public class CompletionRecord {
+
+    @Field
+    private long callingNumber;
```
Comment (larubbio) [inline mobile-academy/src/main/java/org/motechproject/nms/mobileacademy/domain/CompletionRecord.java]: #118 added validation around our storing of calling numbers that you'll need to add here.

    @Min(value = 1000000000L, message = "callingNumber must be 10 digits")
    @Max(value = 9999999999L, message = "callingNumber must be 10 digits")
    @Column(length = 10)

Ideally there would also be some unit tests to validate it.

File: mobile-academy/src/main/java/org/motechproject/nms/mobileacademy/domain/CompletionRecord.java [hunk 2]
```diff
@@ -0,0 +1,69 @@
+package org.motechproject.nms.mobileacademy.domain;
+
+import org.motechproject.mds.annotations.Entity;
+import org.motechproject.mds.annotations.Field;
+
+/**
+ * Tracks the completion record for a given calling number
+ */
+@Entity(tableName = "nms_ma_completion_records")
+public class CompletionRecord {
```
Comment (larubbio) [inline mobile-academy/src/main/java/org/motechproject/nms/mobileacademy/domain/CompletionRecord.java]: extends MdsEntity?

File: mobile-academy/src/main/java/org/motechproject/nms/mobileacademy/domain/CompletionRecord.java [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+    }
+
+    public int getCompletionCount() {
+        return completionCount;
+    }
+
+    public void setCompletionCount(int completionCount) {
+        this.completionCount = completionCount;
+    }
+}
```
Comment (larubbio) [inline mobile-academy/src/main/java/org/motechproject/nms/mobileacademy/domain/CompletionRecord.java]: equals, hashCode and toString?

File: mobile-academy/src/main/java/org/motechproject/nms/mobileacademy/domain/CompletionRecord.java (hunk 2 above)
Comment (koshalt) [inline mobile-academy/src/main/java/org/motechproject/nms/mobileacademy/domain/CompletionRecord.java]: Don't have a need to access any of the auto-gen fields from MDS. Do you still want me to extend it?

File: mobile-academy/src/main/java/org/motechproject/nms/mobileacademy/domain/CompletionRecord.java (hunk 3 above)
Comment (koshalt) [inline mobile-academy/src/main/java/org/motechproject/nms/mobileacademy/domain/CompletionRecord.java]: sync'd offline. Not needed for any of the ma operations.

**Comment by larubbio** [comment]
Other than callingNumber lgtm

File: mobile-academy/src/main/java/org/motechproject/nms/mobileacademy/domain/CompletionRecord.java (hunk 1 above)
Comment (koshalt) [inline mobile-academy/src/main/java/org/motechproject/nms/mobileacademy/domain/CompletionRecord.java]: sounds good. fixing this now.

File: mobile-academy/src/main/java/org/motechproject/nms/mobileacademy/service/impl/MobileAcademyServiceImpl.java [hunk 4]
```diff
@@ -81,30 +109,16 @@ public MaBookmark getBookmark(Long callingNumber, Long callId) {
...
-            } else {
-                toReturn.setScoresByChapter(null);
-                toReturn.setBookmark(null);
-            }
-
+        }
 
-            return toReturn;
+        if (bookmarks.size() > 1) {
+            LOGGER.debug("Found more than 1 instance of valid bookmark, picking top");
```
Comment (frankhuster) [inline mobile-academy/src/main/java/org/motechproject/nms/mobileacademy/service/impl/MobileAcademyServiceImpl.java]: Shouldn't that be LOGGER.warning ?

File: mobile-academy/src/main/java/org/motechproject/nms/mobileacademy/service/impl/MobileAcademyServiceImpl.java [hunk 5]
```diff
@@ -152,4 +176,99 @@ private Bookmark setBookmarkProperties(MaBookmark fromBookmark, Bookmark toBookm
...
+
+        String bookmark = fromBookmark.getChapterIdentifier() + "_" + fromBookmark.getLessonIdentifier();
+        toReturn.setBookmark(bookmark);
+
+        if (fromBookmark.getProgress() != null) {
+
+            Map<String, Integer> scores = getScores(fromBookmark);
+
+            // default behavior to map the data
+            toReturn.setBookmark(bookmark);
```
Comment (frankhuster) [inline mobile-academy/src/main/java/org/motechproject/nms/mobileacademy/service/impl/MobileAcademyServiceImpl.java]: Same as line 186, why have that here?

File: mobile-academy/src/test/java/org/motechproject/nms/mobileacademy/it/MobileAcademyServiceBundleIT.java [hunk 6]
```diff
@@ -122,6 +123,69 @@ public void testSetExistingBookmark() {
...
+
+        MaBookmark retrieved = maService.getBookmark(9987654321L, 666L);
+        assertNotNull(retrieved.getCallingNumber());
+        assertNotNull(retrieved.getCallId());
+        assertNotNull(retrieved.getBookmark());
+        assertNotNull(retrieved.getScoresByChapter());
+    }
+
+    @Test
+    public void testSetGetResetBookmark() {
```
Comment (frankhuster) [inline mobile-academy/src/test/java/org/motechproject/nms/mobileacademy/it/MobileAcademyServiceBundleIT.java]: You're not resetting anything, dude!

Comment (frankhuster) [inline mobile-academy/src/test/java/org/motechproject/nms/mobileacademy/it/MobileAcademyServiceBundleIT.java]: Ah, NVM, I just got it. You set passing scores and check that the bookmark is reset, right?

Comment (koshalt) [inline mobile-academy/src/test/java/org/motechproject/nms/mobileacademy/it/MobileAcademyServiceBundleIT.java]: si si

**Comment by frankhuster** [comment]
lgtm, fix the small stuff in the next commit?

File: mobile-academy/src/main/java/org/motechproject/nms/mobileacademy/service/impl/MobileAcademyServiceImpl.java (hunk 5 above)
Comment (koshalt) [inline mobile-academy/src/main/java/org/motechproject/nms/mobileacademy/service/impl/MobileAcademyServiceImpl.java]: hmm, you're right. I missed this one during the refactor

**Pull request merged**